sigal/plugins/encrypt/encrypt.py

```diff
@@ -0,0 +1,297 @@
+
+Note: The plugin caches the cryptographic parameters (but not the password) after 
+the first build, so that incremental builds can share the same credentials. 
+DO NOT CHANGE THE PASSWORD OR OTHER CRYPTOGRAPHIC PARAMETERS ONCE A GALLERY IS 
+BUILT, or there will be inconsistency in encrypted files and viewers will not be able
+to see some of the images any more.
+'''
+
+import os, random, string, logging, pickle
+from io import BytesIO
+from itertools import chain
+
+from sigal import signals
+from sigal.utils import url_from_path, copy
+from sigal.settings import get_thumb
+from click import progressbar
+from bs4 import BeautifulSoup
+
+from .endec import encrypt, kdf_gen_key
+
+ASSETS_PATH = os.path.normpath(os.path.join(
+    os.path.abspath(os.path.dirname(__file__)), 'static'))
+
+class Abort(Exception):
+    pass
+
+def gen_rand_string(length=16):
+    return "".join(random.SystemRandom().choices(string.ascii_letters + string.digits, k=length))
+
+def get_options(gallery):
+    settings = gallery.settings
+    cache = gallery.encryptCache
+    if "encrypt_options" not in settings:
+        raise ValueError("Encrypt: no options in settings")
+    
+    #try load credential from cache
+    try:
+        options = cache["credentials"]
+    except KeyError:
+        options = settings["encrypt_options"]
+
+    table = str.maketrans({'"': r'\"', '\\': r'\\'})
+    if "password" not in settings["encrypt_options"]:
+        raise ValueError("Encrypt: no password provided")
+    else:
+        options["password"] = settings["encrypt_options"]["password"]
+        options["escaped_password"] = options["password"].translate(table)
+
+    if "gcm_tag" not in options:
+        options["gcm_tag"] = gen_rand_string() 
+    options["escaped_gcm_tag"] = options["gcm_tag"].translate(table)
+    
+    if "kdf_salt" not in options:
+        options["kdf_salt"] = gen_rand_string()
+    options["escaped_kdf_salt"] = options["kdf_salt"].translate(table)
+
+    if "galleryId" not in options:
+        options["galleryId"] = gen_rand_string(6)
+    
+    if "kdf_iters" not in options:
+        options["kdf_iters"] = 10000
+
+    if "ask_password" not in options:
+        options["ask_password"] = settings["encrypt_options"].get("ask_password", False)
+
+    gallery.encryptCache["credentials"] = {
+        "gcm_tag": options["gcm_tag"],
+        "kdf_salt": options["kdf_salt"],
+        "kdf_iters": options["kdf_iters"],
+        "galleryId": options["galleryId"]
+    }
+
+    if "encrypt_symlinked_originals" not in options:
+        options["encrypt_symlinked_originals"] = settings["encrypt_options"].get("encrypt_symlinked_originals", False)
+
+    return options
+
+def cache_key(media):
+    return os.path.join(media.path, media.filename)
+
+def save_property(cache, media):
+    key = cache_key(media)
+    if key not in cache:
+        cache[key] = {}
+        cache[key]["size"] = media.size
+        cache[key]["thumb_size"] = media.thumb_size
```

Comment:
What are those useful for ? (size and thumb_size)

Reply:
They are cached for future incremental builds because once images are encrypted, they are not readable by PIL to get the sizes. Just like how the `extended_caching` plugin works.

Reply:
Also I think the `extended_caching` plugin should be enabled when using `encrypt`, for the same reason. I will update the documentation to reflect that.

Reply:
On further investigation, I think I was wrong about caching exif. Unlike the sizes of the resized images and the thumbnails, the exif info is read from the source images, therefore it is always available: 

https://github.com/saimn/sigal/blob/8547f18e6be2f38485b6edaf89e59726466c4a0c/sigal/gallery.py#L200-L202

https://github.com/saimn/sigal/blob/8547f18e6be2f38485b6edaf89e59726466c4a0c/sigal/gallery.py#L216-L220

So it is not necessary to cache exif data. This issue is considered resolved.

Reply:
Indeed, good point.